dicom_utils/types.py

```diff
@@ -18,34 +18,134 @@
 WINDOW_WIDTH = 0x00281051
 UNKNOWN: Final = -1
 
+
 Dicom = Dataset
 DicomAttributeSequence = Sequence
 
 
+class ModalityError(Exception):
+    pass
+
+
+def get_value(dcm: Dicom, tag: Tag, default: Any) -> Any:
+    return dcm.get(tag).value if tag in dcm else default
+
+
 def get_tag_values(tags: Iterable[Tag], dcm: Dicom) -> Dict[Tag, Any]:
-    tags = {tag: dcm[tag].value if tag in dcm else None for tag in tags}
+    tags = {tag: get_value(dcm, tag, None) for tag in tags}
     return tags
 
 
-class SimpleImageType(Enum):
+# TODO: these would be better in dicom.py, but circular import issues
+def view_code_iterator(dcm: Dataset) -> Iterator[Dataset]:
+    r"""Iterates over all view codes in an input."""
+    view_code_sequence = (get_value(dcm, Tag.ViewCodeSequence, []) or []) if isinstance(dcm, Dataset) else dcm
+    for view_code in view_code_sequence:
+        yield view_code
+
+
+def view_modifier_code_iterator(dcm: Dataset) -> Iterator[Dataset]:
+    r"""Iterates over all view modifier codes in an input."""
+    # view modifier code can be at top level of dicom, or nested in view codes
+    modifier_sequence = (get_value(dcm, Tag.ViewModifierCodeSequence, []) or []) if isinstance(dcm, Dataset) else dcm
+    for modifier in modifier_sequence:
+        yield modifier
+    for view_code in view_code_iterator(dcm):
+        for modifier in view_modifier_code_iterator(view_code):
+            yield modifier
+
+
+T = TypeVar("T")
+
+
+class EnumMixin(Enum):
+    def __bool__(self) -> bool:
+        return not self.is_unknown
+
+    def __repr__(self) -> str:
+        name = self.simple_name
+        return f"{self.__class__.__name__}({name})"
+
+    def __add__(self: T, other: T) -> T:
+        return self or other
+
+    def __mul__(self: T, other: T) -> T:
+        return self or other
+
+    @property
+    def is_unknown(self) -> bool:
+        return self.value == UNKNOWN
+
+    @property
+    def simple_name(self) -> str:
+        return self.name.lower().replace("_", " ")
+
+    @staticmethod
+    def get_required_tags() -> List[Tag]:
+        raise NotImplementedError("get_required_tags() has not been implemented for this class")
+
+
+class MammogramType(EnumMixin):
+    r"""Enum over the subcategories of mammograms"""
     UNKNOWN = 0
-    NORMAL = 1
-    SVIEW = 2
-    TOMO = 3
+    FFDM = 1
+    SFM = 2
+    SVIEW = 3
+    TOMO = 4
 
     @staticmethod
-    def from_dicom(dcm: Dicom) -> "SimpleImageType":
+    def get_required_tags() -> List[Tag]:
+        return [Tag.ImageType, Tag.Modality, Tag.NumberOfFrames, Tag.SeriesDescription]
+
+    @staticmethod
+    def from_dicom(dcm: Dicom, is_sfm: bool = False) -> "MammogramType":
+        if (modality := get_value(dcm, Tag.Modality, None)) not in (None, "MG"):
+            raise ModalityError(f"Expected modality=MG, found {modality}")
```

Comment:
I believe I've seen mammograms with `CT` instead of `MG` as the modality. We may want to check that this assertion generalizes.